### DISCUSSION TITLE
3.8.13 package for RHEL 8 won't start on Erlang 22.0

### DISCUSSION THREAD

**Discussion by assen-totin** [rabbitmq/rabbitmq-server#2856] [Other]
The BinTray repository says it is owned by you:
https://bintray.com/rabbitmq/rpm/rabbitmq-server/3.8.13-1

The RHEL-8 package for 3.8.13 is broken - server refuses to start with 

```
{"Kernel pid terminated",application_controller,"{application_start_failure,credentials_obfuscation,{{shutdown,{failed_to_start_child,credentials_obfuscaton_svc,{undef,[{crypto,macN...
```

The likely reason is here:
https://github.com/rabbitmq/credentials-obfuscation/blob/master/ChangeLog.md#changes-between-230-and-240-february-18-2021

```
The library now supports Erlang 24 and drops support for Erlang versions < 22.1.
```

RHEL-8 only ships Erlang 22.0:

```
[root@localhost assen.totin]# rpm -q erlang
erlang-22.0.7-1.el8.x86_64
```

This needs some attention as the Erlang package in RHEL-8 is not (yet) modular, hence no update of the major or minor version is expected during the lifecycle of the product - and next major RHEL version is still 3 years away.

**Comment by michaelklishin** [comment]
Latest RabbitMQ releases require [Erlang/OTP 22.3 or later](https://www.rabbitmq.com/which-erlang.html). This has been the documented case since `3.8.9`, which shipped in September 2020. So the package is not broken, your system does not provide a supported Erlang version.

Our team produces [zero dependency Erlang RPMs](https://github.com/rabbitmq/erlang-rpm) that go up to `23.2.5` at the moment. Current supported CentOS versions are 7 and 8, and up until relatively recently, even CentOS 6 users could get an Erlang 23 version that RabbitMQ supports.

We do not expect to support Erlang 22 for three more years, in fact, support for it will be dropped as soon as Erlang 24 goes GA this May or June. So our team only supports two most recent major Erlang versions and it has been [the case for a few years](https://www.rabbitmq.com/which-erlang.html).

Please also be aware of the fact that Bintray the service [goes away on May 1st](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/). So perhaps you should consider installing packages from GitHub (an easy option with RPMs) or PackageCloud.

**Reply by assen-totin** [reply]
Every RPM package that installs and then the service refuses to start is broken by definition. You cannot expect sysadmins to read the release notes of each of your bugfix releases - updates are automated nowadays, you know. Not to mention that bumping system requirements in a bugfix release is an awfully bad idea.

**Reply by michaelklishin** [reply]
@assen-totin ah, the good old overgeneralisation, "anything that does X by definition is broken" card.

**TL;DR**: RabbitMQ RPM docs are explicit that the stuff you get from standard distributions is **terribly** out of date. Erlang release support policy of RabbitMQ is widely known and time-based. We always first produce packages for new Erlang versions, then document a release series to be required, then actually enforce it in code 4-6 months later. Those who love auto-updates should use Docker images/OCIs, which always offer the latest versions of dependencies. RHEL software upgrade policy is so laughably out of touch with the rate of changes
of modern tools and dependencies (sometimes critically important ones, e.g. OpenSSL) that no software maintainer should take it seriously.

For starters, [RabbitMQ RPM installation guide](https://www.rabbitmq.com/install-rpm.html) mentions that standard distro dependencies are usually so old they are almost always [out of support](https://www.rabbitmq.com/versions.html) (a typical RabbitMQ release series is supported for 24-30 months).

Erlang releases come out on a schedule, and we support two years worth of release with some rare exceptions (e.g. we cannot support both Erlang 22 and 24, which means 22 will be dropped without a transition period soon). This has been the case for a while. The majority of our community welcomes it. We cannot support an Erlang version for 4-5 years for the rapid rate of crypto API changes alone.

We *always* first documented a release as required and then actually enforce this requirement several months later — in this specific case, 22.3 was documented as a requirement 5 months ago. Erlang 23 has been around for 9 months, and so have our zero dependency Erlang RPMs for it. So you only had to check the release notes every roughly 5-6 months, which is not unreasonable.

If your point is that the Erlang dependency in the RPM spec was not updated, we will take a look. However, RHEL 8 being pinned to a single *minor* release of Erlang for 4-5 years is a problem for RHEL developers to solve, not us.
Given the rate of changes in some key dependencies (programming languages, OpenSSL), this kind of support policy means that RHEL users will soon enough have no choice but to run **everything** in containers, which kind of tells you that the base OS maintainers are doing something terribly wrong. Speaking of containers…

By far the best option for those who love auto upgrades is to use Docker images/OCIs. [RabbitMQ community Docker image](https://github.com/docker-library/rabbitmq) upgrades Erlang, OpenSSL and RabbitMQ to the latest available patch versions within several hours after their release. If you choose to enable automatic image updates, you are always running a very recent versions of all dependencies.

**Reply by michaelklishin** [reply]
I found what the problem was with out RPM spec.

Minimum Erlang version per our RPM package spec is [22.3](https://github.com/rabbitmq/rabbitmq-server-release/blob/master/packaging/RPMS/Fedora/rabbitmq-server.spec#L2). However, only on the `master` branch, not `v3.8.x`.

**Reply by michaelklishin** [reply]
Updated in https://github.com/rabbitmq/rabbitmq-server-release/commit/80faebd01a1589ca7b814010d883b54a9e406b3f. `3.8.14` RPM will in theory refuse to install on Erlang 22.0.

**Reply by McNutnut** [reply]
Doesn't look like the change made it into [3.8.14 release](https://packagecloud.io/rabbitmq/rabbitmq-server/packages/el/8/rabbitmq-server-3.8.14-1.el8.noarch.rpm) as it is still only requiring. `erlang >= 21.3-`.  As a result, installation on RHEL8 is still broken on fresh-install.

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Reply by jikortus** [reply]
Hello @michaelklishin, I ran into the same issue. The point of the whole issue mentioned previously here is that any RPM package should have set its real requirements in the spec file, so as to avoid situations like this, when the package can be installed seemingly without problems, but later on the user finds out that something doesn't work and finding out the underlying root cause (old version of erlang) may not be very obvious. The requirement itself for a newer version of erlang than available in RHEL-8 is fine, but this requirement should be reflected at the RPM level. Otherwise it defeats the basic principles of using a package manager to install packages and is somewhat unexpected for the user, as the dependency error should appear already at the package installation time. Please bear this in mind when creating RPM packages to save the users unnecessary debugging to find out why the rabbitmq-server service fails to start. Thank you!

**Comment by hwangho421** [comment]
@assen-totin
try to disable FIPS mode from RHEL 8.
/usr/bin/fips-mode-setup --check
/usr/bin/fips-mode-setup --disable

seems like there is an issue with FIPS mode on RHEL 8.
